pkg/operator/controller/dns/controller.go

```diff
@@ -0,0 +1,172 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "dns_controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+func New(mgr manager.Manager, operatorNamespace string, dnsProvider dns.Provider) (runtimecontroller.Controller, error) {
+	reconciler := &reconciler{
+		client:            mgr.GetClient(),
+		cache:             mgr.GetCache(),
+		recorder:          mgr.GetEventRecorderFor(controllerName),
+		operatorNamespace: operatorNamespace,
+		dnsProvider:       dnsProvider,
+	}
+	c, err := runtimecontroller.New(controllerName, mgr, runtimecontroller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Watch(&source.Kind{Type: &iov1.DNSRecord{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type reconciler struct {
+	client            client.Client
+	cache             cache.Cache
+	recorder          record.EventRecorder
+	operatorNamespace string
+	dnsProvider       dns.Provider
+	dnsConfig         *configv1.DNS
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// TODO: If the new revision has changed, undefined stuff happens (resource leaks?)
+	dnsConfig := &configv1.DNS{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get dns 'cluster': %v", err)
+	}
+
+	record := &iov1.DNSRecord{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, record); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("dnsrecord not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		} else {
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, fmt.Errorf("failed to get dnsrecord: %v", err)
+		}
+	}
+
+	// If the DNS record was deleted, clean up and return.
+	if record.DeletionTimestamp != nil {
+		if err := r.delete(record); err != nil {
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	ingressName, ok := record.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		return reconcile.Result{}, fmt.Errorf("missing owner label")
+	}
+	ingress := &operatorv1.IngressController{}
+	if err := r.cache.Get(context.TODO(), types.NamespacedName{Namespace: record.Namespace, Name: ingressName}, ingress); err != nil {
+		if errors.IsNotFound(err) {
+			// TODO: what should we do here? something upstream could detect and delete the orphan? add new conditions?
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, fmt.Errorf("ingresscontroller %s not found for record %s", ingressName, record.Name)
+		} else {
+			return reconcile.Result{RequeueAfter: 15 * time.Second}, fmt.Errorf("failed to get ingresscontroller %s for record %s: %v", ingressName, record.Name, err)
```

Comment:
You need a nil error return value in order for `RequeueAfter` to take effect.

Reply:
Phew, good catch! My approach is now to log and retry without returning a reconciler error.